### ISSUE TITLE
[Feature Request] New `BeartypeConf.claw_skip_package_names` configuration option to ignore problematic third-party packages during `beartype.claw` import hooks

### ISSUE THREAD

**Issue by minmax** [beartype/beartype#407]
As discussed in https://github.com/beartype/beartype/issues/404, it will be useful if `beartype.claw.beartype_all` receives an argument `ignore_packages: Iterable[str]` to prevent checking packages with corrupted type hints.

IMHO, `exclude_packages` would be better suited here, because `ignore` may mean handling an error that has occurred.

**Comment by frrad** [comment]
I'm potentially interested in taking this up. Would you mind sketching out what would be involved as you suggested in https://github.com/beartype/pytest-beartype/issues/8#issuecomment-2277224065 ? 

As a disclaimer, if this ends up being too involved, it may be beyond me. I love using a good type checker, but I would likely have a hard time understanding too deep into the internals.

**Comment by leycec** [comment]
> As a disclaimer, if this ends up being too involved, it may be beyond me.

**Absolutely.** Understood. From past experience with ~~willing sacrificial victims~~ wonderful open-source enthusiasts such as yourself, the problem usually isn't the coding aspect. The problem is me. I... can be a bit too much to *bear*. <sup>*hyuk hyuk*</sup>

I tend to type too much – like, *a lot* too much. Frankly, I overwhelm people. When you feel the inevitable desire to just parachute out while you still can, don't feel bad. I expect, welcome, and appreciate the urge to run away. I'd do the same. :smile: 

> Would you mind sketching out what would be involved as you suggested in https://github.com/beartype/pytest-beartype/issues/8#issuecomment-2277224065 ?

**Wonderful!** Yes. It's a two-step process, which easily subdivides this feature request into two distinct pull requests (PRs). Without further ado, <sup>*...what even is an "ado"?* :thinking: </sup> these two PRs are...

## PR 1: Tedious Configuration Stuff So Boring It Blows Minds

In PR 1, please *define a new `BeartypeConf.claw_skip_package_names` configuration option.* This part is easy to understand and mostly just involves a bit of object-oriented boilerplate. Specifically:

* In the private `beartype._conf.confcls` submodule:
   1. Please add a new `from beartype.typing import Iterable` import to the top of this submodule. Thank you! :pray: 
   1. Please add a new `'claw_skip_package_names'` string to the `__slots__` class variable. Thank you! :pray: 
   1. Please add a new `claw_skip_package_names: Iterable[str]` type hint nested in the body of the `    if TYPE_CHECKING:` conditional directly below the `__slots__` class variable. Static type-checkers like mypy and pyright demand this sort of otherwise useless dog vomit. I comply, reluctantly. Thank you! :pray: 
   1. Please add a new `claw_skip_package_names: Iterable[str] = ()` parameter to the existing `__new__()` method. You are already feeling exhausted. You feel your fingers slowly reaching for the parachute. I begin waving goodbye and wish you well on your life's many bold journeys. Thank you! :pray: 
   1. Please add a new `claw_skip_package_names,` item to the existing `conf_args = (` tuple in the body of this `__new__()` method. Somehow, against all odds, you're still here. Your parachute failed to open. You're screaming in terror. I commiserate even as I continue demanding an ever-more absurd laundry list of action items related to this PR. Thank you! :pray: 
   1. Please add a new `claw_skip_package_names=claw_skip_package_names,` key-value pair to the existing `conf_kwargs = dict(` dictionary in the body of this `__new__()` method. You didn't think things could get worse. Then things got worse. Thank you! :pray: 
   1. Please add a new `self._claw_skip_package_names = conf_kwargs['claw_skip_package_names']  # pyright: ignore` assignment in the body of this `__new__()` method. You're now wondering, "Why is this insane? Did this *have* to be this insane? WHAT'S WRONG WITH YOU, MAN!?" And I'm like: "...autism, bro. Autism is always the answer." Thank you! :pray: 
   1. Please add a new read-only public property safely exposing this new configuration operation to external callers: e.g.,

      ```python
      @property
      def claw_skip_package_names(self) -> Iterable[str]:
          return self._claw_skip_package_names
      ```

* In the private `beartype._conf.conftest` submodule:
   1. Please add a new `from collections.abc import Iterable as IterableABC` import to the top of this submodule. Thank you! :pray: 
  1. In the body of the existing `die_if_conf_kwargs_invalid()` function, please add a new `if` conditional validating that the passed value of the `claw_skip_package_names` option actually *is* a valid iterable of strings: e.g.,

      ```python
      # If "claw_skip_package_names" is *NOT* an iterable of strings, raise an
      # exception.
      elif not (
          isinstance(conf_kwargs['claw_skip_package_names'], IterableABC) and
          all(
              isinstance(claw_skip_package_name, str)
              for claw_skip_package_name in conf_kwargs[
                  'claw_skip_package_names']
          )
      ):
          raise BeartypeConfParamException(
              f'Beartype configuration parameter "claw_skip_package_names" '
              f'value {repr(conf_kwargs["claw_skip_package_names"])} not '
              f'iterable of strings.'
          )
      # Else, "claw_skip_package_names" is an iterable of strings.
      ```

* In the test-specific `beartype_test.a00_unit.a40_api.conf.test_confcls` submodule:
  * In the body of the existing `test_conf_dataclass()` unit test, feel free to go ham and just copy-paste the `claw_skip_package_names` option into everything that makes sense. If nothing makes sense anymore, you too now know why @leycec codes alone late on a Friday night from within a smelly shack nestled in the luminous Canadian wilderness. Thank you! :pray: 

And that's just the first PR. Let's stop there, because both of stopped reading pages ago. :joy:

**Comment by frrad** [comment]
Thanks for your very detailed instructions! First PR is up.

**Comment by leycec** [comment]
Thank *you* for your tolerance for pain. Interestingly, the next PR should be *a lot* easier. The actual implementation should only be a one- or two-liner. Most of the "real" work will just be in authoring unit tests that exercise this functionality. As you might imagine, unit-testing import hooks is... *interesting.*

## PR 2: The Future Is So Bright I'm Blinded by the Light

Even more interestingly, there are numerous valid locations throughout the `beartype.claw` package that we could handle this `claw_skip_package_names` blacklist. Although equally valid, some are better than others.

**Wait.** *Oh, boi.* The prior two paragraphs are a bald-faced lie. Previously, I thought this was trivial. Now, I actually thought and realized this is non-trivial. But... fear not! I'll immediately switch over to this on Monday evening and finish up this feature request for you (and everybody else), @frrad.

As a note to my future self, here's more-or-less what I'll need to do tomorrow:

* In the private `beartype.claw._clawstate` submodule:
  * Rename the existing `BeartypeClawState.packages_trie` slot to `packages_trie_whitelist`.
  * Add a new `BeartypeClawState.packages_trie_blacklist` slot of the same `PackagesTrie` type.
  * Initialize `BeartypeClawState.packages_trie_blacklist` to a non-empty `PackagesTrie()` data structure blacklisting *only* the `beartype` package by default.
* In the private `beartype.claw._pkg.clawpkgtrie` submodule:
  * Refactor the existing `iter_packages_trie()` generator function to ignore the passed package name if that name resides in the `claw_state.packages_trie_blacklist` trie.
* Figure out where to communicate the passed `BeartypeConf.claw_skip_package_names` option to the `claw_state.packages_trie_blacklist` trie. (Probably not hard, just to think for a hot minute. *Why's it so hot this summer!?*)

Things are getting spicy in here. :hot_face:

**Comment by leycec** [comment]
*...heh.* Called it. Indeed, this is super-brutal. I couldn't help myself tonight and launched into this full-force, @frrad.

When you're hip deep in [esoteric data structures that nobody's ever heard of and sound like a mobile phone mispelling](https://en.wikipedia.org/wiki/Trie), you know you're in the deep end of it. Thankfully, I should have this wrapped up by Wednesday. At that point, I'll tag @tusharsadhwani in for the glorious `pytest-beartype` finish. Am I the only one sweating in this heat? :sweat_smile:

**Comment by tusharsadhwani** [comment]
Awesome. Wednesday release it is then.

**Comment by leycec** [comment]
![](https://media1.tenor.com/m/XWv-IIBZDrAAAAAC/pump-chest-spongebob.gif)

**Comment by leycec** [comment]
**Resolved by 64622f70d2bf0fe.** This image succinctly describes my feelings about this commit chain:

![](https://media1.tenor.com/m/SUjcWFQri28AAAAC/its-madness-william-frawley.gif)
<sup>*preach, old sad guy.*</sup>

**Comment by tusharsadhwani** [comment]
And I believe that's my cue.
